File: C-Language/Queue/PriorityQueue.c
//Priority Queues with LinkedList

#include <stdio.h>
#include <stdlib.h>

struct QueueNode {
    int data;
    int priority; //Higher the value, higher the priority
    struct QueueNode *next;
} *head = NULL;


struct QueueNode* createQueueNode(int data, int priority) {
    struct QueueNode *newNode = (struct QueueNode*)malloc(sizeof(struct QueueNode));
    newNode->data = data;
    newNode->priority = priority;
    newNode->next = NULL;
    return newNode;
}

void enQueue(int data, int priority) {
    //Insertion at beginning
    struct QueueNode *newNode = createQueueNode(data, priority);
    
    if(head == NULL) {
        //Queue is empty
        head = newNode;
    } else if(head->next == NULL) {
        //Queue contains only one element
        if(newNode->priority <= head->priority) {
            newNode->next = head;
            head = newNode;
        } else {
            head->next = newNode;
        }
    } else {
        //More than one element
        struct QueueNode *current = head;
        struct QueueNode *prev = NULL;
        while(current != NULL && current->priority < newNode->priority) {
            prev = current;
            current = current->next;
        }
        if(prev == NULL) {
            newNode->next = head;
            head = newNode;
        } else {
            prev->next = newNode;
            newNode->next =current;
        }
        
    }
}

void deQueue() {
    //Deletion at the end
    if(head == NULL) {
        //Nothing to delete
        printf("Queue is empty\n");
    } else if (head->next == NULL) {
        //Only one element in queue;
        free(head);
        head = NULL;
    } else {
        //More than one element
        struct QueueNode *current = head;
        struct QueueNode *prev = NULL;
        while(current->next != NULL) {
            prev = current;
            current = current->next;
        }
        prev->next = NULL;
        free(current);
    }
}


void peek() {
    if(head == NULL) {
        //Nothing to delete
        printf("Queue is empty\n");
    } else if (head->next == NULL) {
        //Only one element in queue;
        printf("Peek: %d\n", head->data);
    } else {
        //More than one element
        struct QueueNode *current = head;
        struct QueueNode *prev = NULL;
        while(current->next != NULL) {
            prev = current;
            current = current->next;
        }
        printf("Peek: %d\n", current->data);
    }
}

void display() {
    struct QueueNode *current = head;
    if(current == NULL) {
        printf("Queue is empty\n");
    } else {
        printf("Entry -> ");
        while(current!=NULL) {
            printf("%d -> ", current->data);
            current = current->next;
        }
        printf("Exit\n");
    }
}


int main() {
    printf("Priority Queue with LinkedList\n");
	enQueue(10,1);
	enQueue(20,1);
	enQueue(25,1);
	display(); 
	enQueue(5, 10);
	display(); 
	enQueue(36, 5);
	display(); 
	enQueue(37, 5);
	display(); 
	deQueue();
	display();
	peek();
	
	return 0;
}


/*
Expected Output:
---------------
Priority Queue with LinkedList
Entry -> 25 -> 20 -> 10 -> Exit
Entry -> 25 -> 20 -> 10 -> 5 -> Exit
Entry -> 25 -> 20 -> 10 -> 36 -> 5 -> Exit
Entry -> 25 -> 20 -> 10 -> 37 -> 36 -> 5 -> Exit
Entry -> 25 -> 20 -> 10 -> 37 -> 36 -> Exit
Peek: 36
*/

File: Java/src/ds/queue/PriorityQueue.java
package ds.queue;

public class PriorityQueue {
	private PriorityNode rear;
	private PriorityNode front;
	private int size = 0;
	
	public void enQueue(int data, int priority) {
		PriorityNode newNode = new PriorityNode(data, priority);
		this.size++;
		if(this.rear == null && this.front == null) {
			this.rear = newNode;
			this.front = newNode;
		} else {
			if(newNode.priority <= this.rear.priority) {
				//Insertion at rear
				newNode.next = this.rear;
				this.rear.prev = newNode;
				this.rear = newNode;
			} else if(newNode.priority > this.front.priority) {
				//Insertion at front
				newNode.prev = this.front;
				this.front.next = newNode;
				this.front = newNode;
			} else {
				//Insertion at middle
				PriorityNode curr = rear;
				while(curr.next != null && newNode.priority > curr.priority) {
					curr = curr.next;
				}
				PriorityNode prevNode = curr.prev;
				prevNode.next = newNode;
				newNode.prev = prevNode;
				newNode.next = curr;
				curr.prev = newNode;
				
			}
		}
	}
	
	public int deQueue() {
		if(this.front == null) {
			return -1;
		} else {
			PriorityNode firstNode = this.front;
			PriorityNode prevNode = this.front.prev;
			prevNode.next = null;
			firstNode.prev = null;
			this.front = prevNode;
			return firstNode.data;
		}
	}
	
	
	public int getSize() {
		return this.size;
	}
	
	public boolean isEmpty() {
		return this.size == 0;
	}
	
	public void printQueue() {
		PriorityNode curr = this.rear;
		while(curr!=null) {
			System.out.print(curr.data +  " ");
			curr = curr.next;
		}
		System.out.println();
	}
	
}
